File: Application/src/main/java/com/shareyourproxy/api/rx/command/GetUserMessagesCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;

import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;

import java.util.List;

import static com.shareyourproxy.api.rx.RxMessageSync.getFirebaseMessages;

/**
 * Created by Evan on 6/18/15.
 */
public class GetUserMessagesCommand extends BaseCommand {

    public static final Creator<GetUserMessagesCommand> CREATOR =
        new Creator<GetUserMessagesCommand>() {
            @Override
            public GetUserMessagesCommand createFromParcel(Parcel in) {
                return new GetUserMessagesCommand(in);
            }

            @Override
            public GetUserMessagesCommand[] newArray(int size) {
                return new GetUserMessagesCommand[size];
            }
        };
    private final static java.lang.ClassLoader CL = GetUserMessagesCommand.class.getClassLoader();
    private final String userId;

    public GetUserMessagesCommand(String userId) {
        this.userId = userId;
    }

    private GetUserMessagesCommand(Parcel in) {
        this((String) in.readValue(CL));
    }

    @Override
    public EventCallback execute(Service service) {
        return getFirebaseMessages(service, userId);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(userId);
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/GetUserMessagesCommand.kt
package com.shareyourproxy.api.rx.command

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.rx.RxMessageSync.getFirebaseMessages
import com.shareyourproxy.api.rx.command.eventcallback.UserMessagesDownloadedEventCallback


/**
 * Get user messages.
 */
internal final class GetUserMessagesCommand(private val userId: String) : BaseCommand() {
    @Suppress("UNCHECKED_CAST")
    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as String)

    override fun execute(context: Context): UserMessagesDownloadedEventCallback {
        return getFirebaseMessages(context, userId)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(userId)
    }

    companion object {
        private val CL = GetUserMessagesCommand::class.java.classLoader
        @JvmField val CREATOR: Parcelable.Creator<GetUserMessagesCommand> = object : Parcelable.Creator<GetUserMessagesCommand> {
            override fun createFromParcel(parcel: Parcel): GetUserMessagesCommand {
                return GetUserMessagesCommand(parcel)
            }

            override fun newArray(size: Int): Array<GetUserMessagesCommand?> {
                return arrayOfNulls(size)
            }
        }
    }
}
